Lit rewrite

TheBv opened the pull request:
I saw some work being done on the lit branch again so I decided to whip up this draft before the same work gets done twice.

So far I've ported every component to lit. There are however some challenges which I'm not yet sure how to overcome:

### Parent dependent Styling
Lit uses ShadowDOMs for each component which is causing issues with nested components that used to rely on the parents class for styling. E.g if you define the light-theme for `discord-messages` you can see that the name doesn't have the correct color:
![image](https://user-images.githubusercontent.com/12676257/233682898-91392efc-a08f-488e-a5b5-3a259c37eb54.png)

Similar issues plague the other components as well e.g replies don't render correctly as well. Although some of these components can probably be fixed by adjusting the css directly.

### Icons
We used to style the icons in some components to match the size we needed. TemplateResults which I've used so far based on the previous components can't be styled. 
We can create function driven components but lit doesn't support spreading. So I see two solutions: A: use open-wc's [lit-helpers ](https://open-wc.org/docs/development/lit-helpers/#spread-directivesl) 
or B: We define the function to take the various arguments we need to style the component.

### Notes

Most of the visible issues one can see in the demo pretty much derive from the two issues mentioned above. If anyone has some good ideas/directions on how to tackle them I'd be willing to apply them.

Also I haven't implemented any of the various commits that have happened since the creation of this branch including changes done on the main branch.

There's a lot of cleanup that still needs to be done as well so please keep that in mind :)

favna commented:
I rebased this branch on the base `feat/rewrite-to-lit` and added some other changes as well

favna commented:
So I think I covered the issue with icons in my first reply? That just leaves the issue with parent-dependent styling. There are a few different types I think looking at the old code.

## Scenario 1: `.discord-message .discord-mention`

In this scenario, the styling for `discord-mention` isn't really dependent on its parent. The reason `.discord-message` is there, is to ensure that the styling is scoped to only `discord-mention` that exists within a `discord-message`. So in this case this can just be changed to a single `:host`, as `discord-mention` is always a child of `discord-message`

## Scenario 2: `.discord-light-theme .discord-reaction`

In this scenario, we want to know if the light theme is active or not and if it is applying some style overrides. This scenario is already covered in the rewritten `discord-message` component and doesn't change much.

I think that about covers it? If I'm forgetting a scenario let me know. Quite frankly I think I am forgetting one and it's one I asked help about in the Discord server way back but I never did anything with the info I got... sadly I cannot find it anymore either if so. 

---

Lastly, regarding the CI, yeah I know it fails on the React bundle. All that stuff I still need to figure out how to do. Once all the components are working on the demo page I want to do still do the following in no particular order:
- Migrate the demo to storybook stories
- Write some unit tests for the core src. Nothing all too fancy, primarily to cover different prop combinations that should render different results
- Delete the `old-src` directory, of course.
- Figure out a proper setup of how to publish packages, this includes the React bindings (https://lit.dev/docs/frameworks/react/)

TheBv commented:
Okay I've fixed most issues that I was mentioning now there's a couple more for which I don't know the answer to (or just want to ask about):

1. `DiscordSpoiler`s css assumes a lot of times that `DiscordMessage` is it's parent. Should we just allow this element to only work if `DiscordMessage` is it's parent?
2. `DiscordMessage` styles the `a` and the styling gets lost in child components due to the shadow doms. How should we deal with that? My suggestion would be to export a "LinkStyling" that each component can then add to their own styling. But there probably are better solutions.
3. `DiscordMessage` applies styles extra styling to various sub elements e.g `strong`. This styling also get's lost. I've been able to fix this like [this](https://github.com/TheBv/discord-components/blob/80e8d2c39779293a2350b7371e77430144181671/packages/core/src/components/discord-message/DiscordMessage.ts#L209) although the issue with that solution is that it doesn't work for "normal" `code` elements. Maybe what I suggested in 2. would work here as well?
4. `DiscordMessage` used to have [this](https://github.com/TheBv/discord-components/blob/1a7c4b5880cceb74106a17ebe6af1460e9f59aa7/packages/core/src/components/discord-message/DiscordMessage.ts#L55) styling which didn't work correctly. I moved it to `DiscordMessages` but due to the shadow dom the styling can't be applied correctly. I don't know how to fix this.
5. The badge icon in replies is ever so slightly larger new: 
![image](https://user-images.githubusercontent.com/12676257/233811727-b615d97e-94db-4221-929a-b7e9e8025c25.png)
old: 
![image](https://user-images.githubusercontent.com/12676257/233811715-8c0bd250-a763-4b61-aacf-10f0bfd9df26.png)
I genuinely don't know why. The css seems to be exactly the same.

That about covers everything I've picked up on, I think. I'll focus on implementing all the various changes that happened to the components in the main branch next.

favna commented:
> 1. `DiscordSpoiler`s css assumes a lot of times that `DiscordMessage` is it's parent. Should we just allow this element to only work if `DiscordMessage` is it's parent?

Yes

> 2\. `DiscordMessage` styles the `a` and the styling gets lost in child components due to the shadow doms. How should we deal with that? My suggestion would be to export a "LinkStyling" that each component can then add to their own styling. But there probably are better solutions.

See below at 3

> 3\. `DiscordMessage` applies extra styling to various sub elements e.g `strong`. This styling also get's lost. I've been able to fix this like [this](https://github.com/TheBv/discord-components/blob/80e8d2c39779293a2350b7371e77430144181671/packages/core/src/components/discord-message/DiscordMessage.ts#L209) although the issue with that solution is that it doesn't work for "normal" `code` elements. Maybe what I suggested in 2. would work here as well?

I would rather suggest we force the usage of `discord-inline-code` and regular `code` elements don't get styled the same, so I'd say the current behaviour is desired.

Actually, on that note we should also do `discord-link` and not style regular `a` tags. Good thing is is a major version so we can break it to hell and back hehe.

For the rest I'm going to check out the PR again

favna commented:
> 5\. The badge icon in replies is ever so slightly larger new:
>     ![image](https://user-images.githubusercontent.com/12676257/233811727-b615d97e-94db-4221-929a-b7e9e8025c25.png)
>     old:
>     ![image](https://user-images.githubusercontent.com/12676257/233811715-8c0bd250-a763-4b61-aacf-10f0bfd9df26.png)
>     I genuinely don't know why. The css seems to be exactly the same.

I just checked this through the dev tools and both with the new and the old it reports at 15x15 pixels. Switching between tabs also seems identical.

favna commented:
A few things I notice when switching between tabs of the "No framework" example on codesandbox (use "open in new window" in codesandbox for a full view) and running `yarn start` in this branch:

1. The timestamp was closer to the username before
2. The orange of a mention doesn't show
3. Compact mode paddings/margins seem to be wrong


<details>
<summary>Screenshots for 1 and 2</summary>

**New:**
![image](https://user-images.githubusercontent.com/4019718/236665613-28bc1b3f-00a3-4235-b0c8-77697f9485c8.png)

**Old:**
![image](https://user-images.githubusercontent.com/4019718/236665620-36cca019-6383-4484-a552-13a236192dbe.png)

</details>

<details>
<summary>Screenshot for 3</summary>

**New:**
<img width="642" alt="image" src="https://user-images.githubusercontent.com/4019718/236665708-0a21f6ff-c59f-4526-b347-43256094ddc1.png">


**Old:**
<img width="420" alt="image" src="https://user-images.githubusercontent.com/4019718/236665674-8446e5f8-2057-41a9-b6f6-321810f4e7c2.png">

</details>


> edit: One thing I forgot to mention was that the built in styling in the embed example doesn't work as well. I assume we need to import the styles to make these things work? I'm not sure how to do that though.

Regarding this, I assume you're referring to the very weird styling here:
<img width="595" alt="image" src="https://user-images.githubusercontent.com/4019718/236667564-71640839-4af8-4d76-b009-3735150c6d8c.png">

That is indeed peculiar. I'm also not sure how to fix it. Styles are quite all over the place there in the devtools inspector.

In general though I can say that we should use `:host` much more as opposed to wrapping the component in a single `div` for a class name, that has caused various issues such as 6db928b25de380d4cb82c8066b0f727bebcb5031 and ab993b06c145eee40a895445c01d7a32a13c8c17

TheBv commented:
Okay I'm pretty certain I've addressed all the issues that you mentioned. 
I also went ahead and added a `DiscordPre` Element and generified `DiscordCode` so it supports multiline code blocks as well light theming and embed support.

One caveat: Elements that are styled differently in an embed (e.g. DiscordCode) don't get the embed styling automatically. 
The only way to achieve that would be to iterate through all of the elements parents until it finds one that is an embed component, or lightThemed. I decided not to do that though so the user has to define those properties themselves. One could add them later though.

Also should we rename `DiscordCustomEmoji` to `DiscordEmoji` I feel like it'd make more sense given our current naming convention.

favna commented on packages/core/src/components/discord-embed/DiscordEmbed.ts:
```diff
@@ -0,0 +1,402 @@
+	@property()
+	public provider: string;
+
+	// TODO: private hasPerformedRerenderChecks: 'dirty' | 'pristine' = 'pristine';
```
Putting a comment on this line so we don't forget. Ref: https://github.com/TheBv/discord-components/blob/8320ddfac3cc3390fe08b3f7a7c151eeed3861bd/packages/core/src/components/discord-embed/discord-embed.tsx#L91-L100

favna commented on packages/core/src/options.ts:
```diff
@@ -44,16 +44,18 @@ export interface Emoji {
 }
 
-const globalAvatars: Avatars = window.$discordMessage?.avatars ?? ({} as Avatars);
+const serverSideSafeWindow = window ?? globalThis;
```
@TheBv I hope this fixes the NextJS issue but can you confirm?

TheBv replied:
Tested it. We need to check for `typeof window !== undefined` though because that's the only way next.js can properly check if window is defined or not (due to some server-side/client-side stuff).

favna reviewed (changes requested):
There are some bugs left with light mode, for example mentions don't colour correctly:
<img width="866" alt="image" src="https://github.com/skyra-project/discord-components/assets/4019718/dc2c57aa-8f04-4a0b-b90b-33dc133700ab">

favna commented on packages/core/src/options.ts:
```diff
@@ -44,18 +44,21 @@ export interface Emoji {
 
-const serverSideSafeWindow = window ?? globalThis;
+// We need to test for typeof window !== 'undefined' because of how most SSR frameworks work.
+const serverSideSafeWindow = typeof window !== 'undefined' && window !== undefined && window !== null ? window : globalThis;
```
```suggestion
const serverSideSafeWindow = typeof window === 'undefined' || !window ? globalThis : window;
```

Slight adjustment to avoid [no-negated-condition](https://eslint.org/docs/latest/rules/no-negated-condition) pattern

TheBv commented on packages/core/src/components/discord-messages/DiscordMessages.ts:
https://github.com/TheBv/discord-components/blob/b7b0b633fefb42b91edaf76c6aeab6e6e1d4919e/packages/core/src/components/discord-messages/DiscordMessages.ts#L68
What exactly are we trying to check for on connectCallback()?
I feel like the if statements right now can't be correct.

favna replied:
Huh.... uhm... looking at main I ask myself the same thing.

It should just be
```ts
if (defaultTheme === 'light' && (this.lightTheme === undefined || this.lightTheme === null)) {
	this.lightTheme = true;
}

if (defaultMode === 'compact' && (this.compactMode === undefined || this.compactMode === null)) {
	this.compactMode = true;
}

if (defaultBackground === 'none' && (this.defaultBackground === undefined || this.defaultBackground === null)) {
	this.noBackground = true;
}
```

This way if the default is set to whatever `true` would be _and_ the value wasn't already set by the user through props, then we set it to true.

favna replied:
Adjusted

favna commented on packages/core/src/components/discord-embed-field/DiscordEmbedField.ts:
```diff
@@ -0,0 +1,105 @@
+
+		const emojiParsedEmbedFieldTitle = this.parseTitle(this.fieldTitle);
+
+		return html` ${emojiParsedEmbedFieldTitle ? html`<div class="discord-field-title">${[...emojiParsedEmbedFieldTitle]}</div>` : null}
```
```suggestion
		return html`${emojiParsedEmbedFieldTitle ? html`<div class="discord-field-title">${[...emojiParsedEmbedFieldTitle]}</div>` : null}
```

favna commented on packages/core/src/components/discord-embed-footer/DiscordEmbedFooter.ts:
```diff
@@ -0,0 +1,86 @@
+	protected override render() {
+		this.updateTimestamp(this.timestamp);
+
+		return html` ${this.footerImage ? html`<img src="${this.footerImage}" alt="${this.footerImageAlt}" class="discord-footer-image" />` : null}
```
```suggestion
		return html`${this.footerImage ? html`<img src="${this.footerImage}" alt="${this.footerImageAlt}" class="discord-footer-image" />` : null}
```

favna commented:
Something weird is happening with ephemeral replies. I can't quite find out what and why:
![image](https://github.com/skyra-project/discord-components/assets/4019718/315635cd-1342-4f04-b32f-7e19c5b4ff2e)

Also in **non** ephemeral with compact mode the text doesn't center out correctly with the username:
![image](https://github.com/skyra-project/discord-components/assets/4019718/c3cd93bd-9cbc-4e60-b94d-805bde8ddd3c)

favna reviewed (changes requested):
Some things I am noticing that I'd just like to drop here:
1. Kinda mentioned it before but there are some wacky things with compact mode. Outlining, font sizes, etc.
2. The icon for channel mentions isn't properly aligned. 
![image](https://github.com/skyra-project/discord-components/assets/4019718/6e124d0a-7931-4d36-a3ab-ed259ea36c57). I can't tell for sure myself but I also think the other mention SVGs are slightly misaligned.

3. We have 3 more channel types for mentions to add:
![image](https://github.com/skyra-project/discord-components/assets/4019718/10070e01-637c-49e7-8a76-84f2ac7b1e96)

SVG extracted from Discord client:
**Server Guide**
```html
<svg class="icon-2Ph-Lv" aria-label="Server Guide" aria-hidden="false" role="img" width="24" height="24" viewBox="0 0 16 16" fill="none"><path fill-rule="evenodd" clip-rule="evenodd" d="M7 15H9V8H12.5L15 5.5L12.5 3H9V1H7V3H1L3.5 5.5L1 8H7V15Z" fill="currentColor"></path><path d="M5 14C5 12.8954 5.89543 12 7 12H9C10.1046 12 11 12.8954 11 14V15H5V14Z" fill="currentColor"></path></svg>
```

**Browse Channels**
```html
<svg class="icon-2Ph-Lv" aria-label="Browse Channels" aria-hidden="false" role="img" width="24" height="24" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M5.00328 17.5C4.69218 17.5 4.4566 17.2189 4.51097 16.9126L4.99838 14.1667H2.26014C1.94953 14.1667 1.71409 13.8864 1.76763 13.5805L1.8843 12.9138C1.92617 12.6746 2.13392 12.5 2.37681 12.5H5.29005L6.17338 7.5H3.43514C3.12453 7.5 2.88909 7.21977 2.94263 6.91381L3.0593 6.24714C3.10117 6.00789 3.30892 5.83333 3.55181 5.83333H6.46505L6.98347 2.91262C7.02584 2.67391 7.23335 2.5 7.47578 2.5H8.12681C8.43792 2.5 8.67349 2.78107 8.61912 3.08738L8.13171 5.83333H13.1317L13.6501 2.91262C13.6925 2.67391 13.9 2.5 14.1424 2.5H14.7935C15.1046 2.5 15.3402 2.78107 15.2858 3.08738L14.7984 5.83333H17.5366C17.8472 5.83333 18.0827 6.11356 18.0291 6.41952L17.9125 7.08619C17.8706 7.32544 17.6628 7.5 17.4199 7.5H14.5067L14.2417 9H12.5754L12.8404 7.5H7.84041L6.95708 12.5H8.99835V14.1667H6.66505L6.14662 17.0874C6.10425 17.3261 5.89675 17.5 5.65431 17.5H5.00328Z" fill="currentColor"></path><path fill-rule="evenodd" clip-rule="evenodd" d="M10.4 14C10.4 12.0118 12.0118 10.4 14 10.4C15.9883 10.4 17.6 12.0118 17.6 14C17.6 14.7775 17.3536 15.4975 16.9345 16.0859L18.9243 18.0758C19.1586 18.3101 19.1586 18.69 18.9243 18.9243C18.69 19.1586 18.3101 19.1586 18.0758 18.9243L16.0859 16.9345C15.4975 17.3536 14.7775 17.6 14 17.6C12.0118 17.6 10.4 15.9883 10.4 14ZM14 11.6C12.6745 11.6 11.6 12.6745 11.6 14C11.6 15.3255 12.6745 16.4 14 16.4C15.3255 16.4 16.4 15.3255 16.4 14C16.4 12.6745 15.3255 11.6 14 11.6Z" fill="currentColor"></path></svg>
```

**Customize Community**
```html
<svg aria-label="Customise Community" aria-hidden="false" role="img" class="icon-2Ph-Lv" width="24" height="24" viewBox="0 0 24 24"><path d="M21.487 5.126L12.487 0.126C12.184 -0.042 11.818 -0.042 11.515 0.126L2.51498 5.126C2.19798 5.302 2.00098 5.636 2.00098 5.999C2.00098 6.693 2.11498 22.999 12.001 22.999C21.887 22.999 22.001 6.693 22.001 5.999C22.001 5.636 21.804 5.302 21.487 5.126ZM12.001 5.999C13.382 5.999 14.501 7.118 14.501 8.499C14.501 9.88 13.382 10.999 12.001 10.999C10.62 10.999 9.50098 9.88 9.50098 8.499C9.50098 7.118 10.62 5.999 12.001 5.999ZM8.25098 16C8.25098 13.699 9.69998 12.25 12.001 12.25C14.302 12.25 15.751 13.699 15.751 16H8.25098Z" fill="currentColor"></path></svg>
```

4. `DiscordLink` doesn't use the consumer for lightTheme yet

Will probably edit this message as I progress further writing unit tests

favna commented:
@TheBv I forced your branch changes onto `feat/rewrite-to-lit` because it was a merge conflict hellish clusterfuck to do it properly so your PR will show up as merged but ofc in reality it's not yet done, so we'll have to continue on a new PR when you got time. I have also created https://github.com/skyra-project/discord-components/pull/367 as placeholder PR for tracking what will go into the `main` branch.
